File: Python/Leetcode Easy/543DiameterOfBinaryTree.py
#https://leetcode.com/problems/diameter-of-binary-tree/description/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.res = 0
        def helper(node):
            if node == None: return 0
            left = helper(node.left)
            right = helper(node.right)
            self.res = max(self.res, left + right)
            return 1 + max(left, right)
        helper(root)
        return self.res

File: C++/Leetcode Easy/543DiameterOfBinaryTree.cpp
#include <algorithm>
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res;

    int diameterOfBinaryTree(TreeNode* root) {
        res = 0;
        helper(root);
        return res;
    }

private:
    int helper(TreeNode* root){
        if (!root) return 0;
        int left = helper(root->left);
        int right = helper(root->right);
        res = max(res, left+right);
        return 1 + max(left, right);
    }
};


class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int res = 0;
        helper(root, res);
        return res;
    }

private:
    int helper(TreeNode* root, int& res){
        if (!root) return 0;
        int left = helper(root->left, res);
        int right = helper(root->right, res);
        res = max(res, left+right);
        return 1 + max(left, right);
    }
};